infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/RoutingQueuesSnapshotReader.java

```diff
@@ -26,4 +48,76 @@ public void reset(long ts) {
         // In addition to setting the snapshot timestamp to ts, write to the table subscribed to by the client
         // requesting for a snapshot sync
     }
+
+    @Override
+    public SnapshotReadMessage read(UUID syncRequestId) {
+        List<LogReplication.LogReplicationEntryMsg> messages = new ArrayList<>();
+
+        boolean endSnapshotSync = false;
+        LogReplication.LogReplicationEntryMsg msg;
+
+        currentStreamInfo = new OpaqueStreamIterator(streamToReplicate, rt, snapshotTimestamp);
+        log.info("Start Snapshot Sync replication for stream name={}, id={}", streamToReplicate,
+            CorfuRuntime.getStreamID(streamToReplicate));
+
+
+        if (currentStreamHasNext()) {
+            msg = read(currentStreamInfo, syncRequestId);
+            if (msg != null) {
+                messages.add(msg);
+            }
+        } else {
+            endSnapshotSync = true;
+        }
+        return new SnapshotReadMessage(messages, endSnapshotSync);
+    }
+
+    @Override
+    protected boolean currentStreamHasNext() {
+
+        long start = System.currentTimeMillis();
+        long end = start + DATA_WAIT_TIMEOUT_MS;
+
+        while (start < end) {
+            if (super.currentStreamHasNext()) {
+                // Get the entry
+                OpaqueEntry opaqueEntry = (OpaqueEntry) currentStreamInfo.iterator.next();
```

Comment:
since here its just 1 stream, what if we use a stream listener instead of an iterator?

Reply:
A stream listener will give updates starting from a particular timestamp.  In this case, we require all updates upto a certain timestamp.

Also, relying on the stream listener will add delay in receiving updates.  It may be the reason why it was not considered for LogEntryReader for the FULL_TABLE replication model.

I am open for discussion on this.